src/apps/angle/ethereum/angle.balance-fetcher.ts

```diff
@@ -0,0 +1,214 @@
+import { Inject } from '@nestjs/common';
+import { utils, BigNumber } from 'ethers';
+import { compact, sumBy } from 'lodash';
+
+import { drillBalance } from '~app-toolkit';
+import { IAppToolkit, APP_TOOLKIT } from '~app-toolkit/app-toolkit.interface';
+import { Register } from '~app-toolkit/decorators';
+import { presentBalanceFetcherResponse } from '~app-toolkit/helpers/presentation/balance-fetcher-response.present';
+import { BalanceFetcher } from '~balance/balance-fetcher.interface';
+import { isBorrowed, isSupplied } from '~position/position.utils';
+import { Network } from '~types/network.interface';
+
+import { ANGLE_DEFINITION } from '../angle.definition';
+import { AngleContractFactory } from '../contracts';
+import { callAngleApi } from '../helpers/angle.api';
+
+const network = Network.ETHEREUM_MAINNET;
+
+type TVault = {
+  address: string;
+  collateral: string;
+  collateralAmount: number;
+  collateralRatio: number;
+  debt: number;
+  debtString: string;
+  id: number;
+  liquidationPrice: number;
+  rate: number;
+  stablecoin: string;
+  symbol: string;
+};
+
+type TPerpetual = {
+  perpetualID: string;
+  owner: string;
+  decimals: string;
+  margin: string;
+  committedAmount: string;
+  entryRate: string;
+  perpetualManager: string;
+  stableAddress: string;
+  collatAddress: string;
+  stableName: string;
+  collatName: string;
+  openingTimestamp: string;
+  lastUpdateTimestamp: string;
+  openingBlockNumber: string;
+  lastUpdateBlockNumber: string;
+  status: string;
+  claimable: number;
+};
+
+@Register.BalanceFetcher(ANGLE_DEFINITION.id, network)
+export class EthereumAngleBalanceFetcher implements BalanceFetcher {
+  constructor(
+    @Inject(APP_TOOLKIT) private readonly appToolkit: IAppToolkit,
+    @Inject(AngleContractFactory) private readonly angleContractFactory: AngleContractFactory,
+  ) {}
+
+  async getVeAngleTokenBalances(address: string) {
+    return this.appToolkit.helpers.tokenBalanceHelper.getTokenBalances({
+      address,
+      appId: ANGLE_DEFINITION.id,
+      groupId: ANGLE_DEFINITION.groups.veangle.id,
+      network: Network.ETHEREUM_MAINNET,
+    });
+  }
+
+  async getSanUSDCClaimableBalance(address: string) {
+    const { rewardsData } = await callAngleApi<{ rewardsData: { totalClaimable: number } }>('dao', {
+      chainId: 1,
+      user: address,
+    });
```

Comment:
Can you get this from an on-chain request? This will be extremely taxing for your API (our traffic is large).

Reply:
added a cache. i set ttl between 5 and 60 minutes depending on calls. This can probably be set higher. Let me know if that works and if you want me to change something @immasandwich